junit-platform-engine/src/main/java/org/junit/platform/engine/TestDescriptor.java

```diff
@@ -172,6 +176,40 @@ default Set<? extends TestDescriptor> getDescendants() {
 	 */
 	void removeFromHierarchy();
 
+	/**
+	 * Order all children from this descriptor.
+	 *
+	 * <p>The {@code orderer} is provided a modifiable list of child test descriptors in
+	 * this test descriptor. Never {@code null}. The {@code orderer} must return
+	 * a list of descriptors in any order, never {@code null}.
+	 * <p>
+	 * If descriptors were added, these descriptors will be ignored. If descriptors
+	 * are removed, the original descriptors will be retained in arbitrary order.
+	 *
+	 * @param orderer a unary operator to order the children from this test descriptor.
+	 */
+	@API(since = "5.12", status = EXPERIMENTAL)
+	default void orderChildren(UnaryOperator<List<TestDescriptor>> orderer) {
+		Preconditions.notNull(orderer, "orderer must not be null");
+		Set<? extends TestDescriptor> originalChildren = getChildren();
+		List<TestDescriptor> suggestedOrder = orderer.apply(new ArrayList<>(originalChildren));
+		Preconditions.notNull(suggestedOrder, "orderer may not return null");
+
+		Set<? extends TestDescriptor> orderedChildren = new LinkedHashSet<>(suggestedOrder);
+		boolean unmodified = originalChildren.equals(orderedChildren);
+		Preconditions.condition(unmodified, "orderer may not add or remove test descriptors");
+		Preconditions.condition(originalChildren.size() == suggestedOrder.size(),
+			"orderer may not add duplicate test descriptors");
```

Comment:
I feel we could express both these requirements more clearly and in a single assertion. `orderer may only change the ordering of test descriptors`.

Reply:
I think your original error message `orderer may not add or remove test descriptors` is clearer and have merged these two lines in 56dfbfb5ed79f59dad86b25bb0a3090ad77d6035.